nilearn/decoding/space_net.py

```diff
@@ -0,0 +1,1231 @@
+
+    solver_params: dict
+       Dictionary of param-value pairs to be passed to solver.
+    """
+
+    if l1_ratios is None:
+        raise ValueError("l1_ratios must be specified!")
+
+    # make local copy of mask
+    mask = mask.copy()
+
+    # misc
+    _, n_features = X.shape
+    verbose = int(verbose if verbose is not None else 0)
+
+    # Univariate feature screening. Note that if we have only as few as 100
+    # features in the mask's support, then we should use all of them to
+    # learn the model i.e disable this screening)
+    do_screening = (n_features > 100) and screening_percentile < 100.
+    if do_screening:
+        X, mask, support = _univariate_feature_screening(
+            X, y, mask, is_classif, screening_percentile)
+
+    # crop the mask to have a tighter bounding box
+    mask = _crop_mask(mask)
+
+    # get train and test data
+    X_train, y_train = X[train].copy(), y[train].copy()
+    X_test, y_test = X[test].copy(), y[test].copy()
+
+    # it is essential to center the data in regression
+    X_train, y_train, _, y_train_mean, _ = center_data(
+        X_train, y_train, fit_intercept=True, normalize=False,
+        copy=False)
+
+    # misc
+    if isinstance(l1_ratios, numbers.Number):
+        l1_ratios = [l1_ratios]
+    l1_ratios = sorted(l1_ratios)[::-1]  # from large to small l1_ratios
+    best_score = -np.inf
+    best_secondary_score = -np.inf
+    best_l1_ratio = l1_ratios[0]
+    best_alpha = None
+    best_init = None
+    all_test_scores = []
+    if len(test) > 0.:
+        # do l1_ratio path
+        for l1_ratio in l1_ratios:
+            this_test_scores = []
+
+            # make alpha grid
+            if alphas is None:
+                alphas_ = _space_net_alpha_grid(
+                    X_train, y_train, l1_ratio=l1_ratio, eps=eps,
+                    n_alphas=n_alphas, logistic=is_classif)
+            else:
+                alphas_ = alphas
+            alphas_ = sorted(alphas_)[::-1]  # from large to small l1_ratios
+
+            # do alpha path
+            if best_alpha is None:
+                best_alpha = alphas_[0]
+            init = None
+            for alpha in alphas_:
+                # setup callback mechanism for early stopping
+                early_stopper = EarlyStoppingCallback(
+                    X_test, y_test, is_classif=is_classif, debias=debias,
+                    verbose=verbose)
+                w, _, init = solver(
+                    X_train, y_train, alpha, l1_ratio, mask=mask, init=init,
+                    callback=early_stopper, verbose=max(verbose - 1, 0.),
+                    **solver_params)
+
+                # We use 2 scores for model selection: the second one is to
+                # disambiguate between regions of equivalent spearman
+                # correlations
+                score, secondary_score = early_stopper.test_score(w)
+                this_test_scores.append(score)
+                if (np.isfinite(score) and
```

Comment:
Division by zero 

Sent from my phone. Please forgive brevity and mis spelling



On Jul 15, 2015, 21:44, at 21:44, bthirion <notifications@github.com> wrote:
>> +            for alpha in alphas_:
>> +                # setup callback mechanism for early stopping
>> +                early_stopper = EarlyStoppingCallback(
>> +                    X_test, y_test, is_classif=is_classif,
>debias=debias,
>> +                    verbose=verbose)
>> +                w, _, init = solver(
>> +                    X_train, y_train, alpha, l1_ratio, mask=mask,
>init=init,
>> +                    callback=early_stopper, verbose=max(verbose - 1,
>0.),
>> +                    **solver_params)
>> +
>> +                # We use 2 scores for model selection: the second
>one is to
>> +                # disambiguate between regions of equivalent
>spearman
>> +                # correlations
>> +                score, secondary_score = early_stopper.test_score(w)
>> +                this_test_scores.append(score)
>> +                if (np.isfinite(score) and
>
>Just curious: how could a correlation score be infinite ?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nilearn/nilearn/pull/657/files#r34720497